addons/account/models/account_account.py

```diff
@@ -999,37 +997,81 @@ def write(self, vals):
-        for account in accounts:
-            for company in account.company_ids:
+        """ Check account codes per companies. These are the checks:
+
+            For each `account` in self,
+            - for each `company` in `account.company_ids`, `account.with_company(company).code` must be set.
+              This can be checked by checking that the code is set for each company in account.company_ids.root_id.
+
+            - for each `company` in `account.company_ids`, there must not be any `other_account` that is:
+                - accessible from a same company as `account`; and
+                - has the same code as `account` in a company from which both are accessible.
+
+              Noting that companies can access all the accounts of their parents, we can check this by ensuring that there
+              is no `other_account` that belongs to a parent of `account.company_ids` or a child of `account.company_ids`,
+              and that have the same code in the root company.
+
+            For example, if account.company_ids == res.company(1), then, we check:
+            - account.with_company(1).code != False
+            - self.with_company(res.company(1).root_id).search([
+                ('company_ids', 'in' <parents and children of res.company(1)>),
+                ('code', '=', account.with_company(1).code),
+              ]) must be empty.
+        """
+        # Check 1: Check that the code is set.
+        for account in self.sudo():
+            for company in account.company_ids.root_id:
                 if not account.with_company(company).code:
                     raise ValidationError(_("The code must be set for every company to which this account belongs."))
-        accounts_to_check = accounts.filtered(lambda a: a.code and self.env.company in a.company_ids)
-        accounts_by_code = accounts_to_check.grouped('code')
-        duplicate_codes = None
-        if len(accounts_by_code) < len(accounts_to_check):
-            duplicate_codes = [code for code, accounts in accounts_by_code.items() if len(accounts) > 1]
-        # search for duplicates of self in database
-        elif duplicates := self.sudo().search_fetch(
-            [
-                ('code', 'in', list(accounts_by_code)),
-                ('id', 'not in', self.ids),
-            ],
-            ['code'],
-        ):
-            duplicate_codes = duplicates.mapped('code')
-        if duplicate_codes:
-            raise ValidationError(
-                _("Account codes must be unique. You can't create accounts with these duplicate codes: %s", ", ".join(duplicate_codes))
-            )
+
+        # Do a grouping by companies in `company_ids`.
+        account_ids_to_check_by_company = defaultdict(list)
+        for account in self.sudo():
+            companies_to_check = account.company_ids
+            for company in companies_to_check:
+                account_ids_to_check_by_company[company].append(account.id)
```

Comment:
Does it matter where in the branch we are? Can't you just group on the root company?
```suggestion
        accounts_to_check_by_company = self.sudo().grouped('company_ids.root_id')
```